### PR TITLE
Add the ability to specify tags for new written data

### LINKED ISSUE #1460: Add the ability to specify tags for new written data
Currently, there is no way to specify tags for a new written asset. This issue deals with adding a way to do so.

As suggested by @shlomitk1 fybrikapplication can be extended to include resourceMetadata in the flow requirements that the user will fill, for example:
```
    - dataSetID: 'new-data-parquet'
      flow: write
      requirements:
        flowParams:
          isNewDataSet: true
          catalog: fybrik-notebook-sample
          metadata:
            tags:
              finance: true
            columns:
            - name: nameOrig
              tags:
                PII: true
            - name: oldbalanceOrg
              tags:
                sensitive: true
```

blocked by issue #1458

### PR DESCRIPTION
**Pull request by revit13**
Closes #1460 

Currently, there is no way to specify tags for a new written asset. This issue deals with adding a way to do so.

As suggested by @shlomitk1 fybrikapplication can be extended to include resourceMetadata in the flow requirements that the user will fill, for example:

```
 - dataSetID: 'new-data-parquet'
      flow: write
      requirements:
        flowParams:
          isNewDataSet: true
          catalog: fybrik-notebook-sample
          metadata:
            tags:
              finance: true
            columns:
            - name: nameOrig
              tags:
                PII: true
            - name: oldbalanceOrg
              tags:
                sensitive: true
```

Signed-off-by: Revital Sur <eres@il.ibm.com>
Co-authored-by: mohammad-nassar10 <mohammad.nassar@ibm.com>

### CODE REVIEW

File: manager/controllers/app/data_catalog_interface.go [hunk 1]
```diff
@@ -30,18 +32,23 @@ func (r *FybrikApplicationReconciler) RegisterAsset(assetID string, catalogID st
 	}
 
 	var resourceMetadata datacatalog.ResourceMetadata
-	if info.ResourceMetadata != nil {
-		resourceMetadata = *info.ResourceMetadata.DeepCopy()
+	if resourceMetadataFromRequirements != nil {
+		resourceMetadata = *resourceMetadataFromRequirements.DeepCopy()
```
Comment (shlomitk1) [inline manager/controllers/app/data_catalog_interface.go:36]: DeepCopy is applied twice for the same object

**Comment by shlomitk1** [comment]
I have a problem with using metadata that is not in catalog yet for governance decisions. 
Suppose I want to update someone else's asset which I am not allowed to do since I am not the owner. But I could provide my name as an owner in a new metadata, and suddenly it's ok, I have overwritten the asset marking myself as an owner.

**Comment by roytman** [comment]
> I have a problem with using metadata that is not in catalog yet for governance decisions. Suppose I want to update someone else's asset which I am not allowed to do since I am not the owner. But I could provide my name as an owner in a new metadata, and suddenly it's ok, I have overwritten the asset marking myself as an owner.

Can we check that an asset doesn't exist before creating a new catalog entry for it?

File: manager/controllers/app/data_catalog_interface.go [hunk 2]
```diff
@@ -16,11 +16,13 @@ import (
...
 // - assetID: DataSetID as it appears in fybrik-application
 // - catalogID: the destination catalog identifier
+// - resourceMetadataFromRequirements: resource metadata from the flow requirements
 // - info: connection and credential details
 // Returns:
 // - an error if happened
 // - the new asset identifier
-func (r *FybrikApplicationReconciler) RegisterAsset(assetID string, catalogID string, info *app.DatasetDetails,
+func (r *FybrikApplicationReconciler) RegisterAsset(assetID string, catalogID string,
+	resourceMetadataFromRequirements *datacatalog.ResourceMetadata, info *app.DatasetDetails,
```
[the remark this answers is not in the document]
Comment (revit13) [reply manager/controllers/app/data_catalog_interface.go:25]: I find that merging the resource metadata is overkill as in `provisionedBucketRef.ResourceMetadata` holds only information about the geography. I will try to check conflicts only in that field.

**Comment by revit13** [comment]
@roytman the resource metadata from the application flow requirements can be used only when 'isNewDataSet` is true so we will not update an existing asset. In the current implementation if `isNewDataSet` is true then a new asset is registered in the catalog - it is not updating an existing asset

File: manager/controllers/app/fybrikapplication_controller.go [hunk 3]
```diff
@@ -542,23 +546,26 @@ func (r *FybrikApplicationReconciler) checkGovernanceActions(configEvaluatorInpu
...
 		}
 	case taxonomy.ReadFlow, taxonomy.DeleteFlow:
 		reqAction := policymanager.RequestAction{
 			ActionType:         configEvaluatorInput.Request.Usage,
 			Destination:        configEvaluatorInput.Workload.Cluster.Metadata.Region,
 			ProcessingLocation: taxonomy.ProcessingLocation(configEvaluatorInput.Workload.Cluster.Metadata.Region),
 		}
-		req.Actions, err = LookupPolicyDecisions(req.Context.DataSetID, r.PolicyManager, appContext, &reqAction)
+		req.Actions, err = LookupPolicyDecisions(req.Context.DataSetID, req.Context.Requirements.FlowParams.ResourceMetadata,
+			r.PolicyManager, appContext, &reqAction)
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:568]: What is the meaning of taking metadata not from the catalog for read/delete scenarios?

File: manager/controllers/app/fybrikapplication_controller.go [hunk 4]
```diff
@@ -542,23 +546,26 @@ func (r *FybrikApplicationReconciler) checkGovernanceActions(configEvaluatorInpu
 	var err error
 	switch configEvaluatorInput.Request.Usage {
 	case taxonomy.WriteFlow:
-		if !req.Context.Requirements.FlowParams.IsNewDataSet {
+		if !req.Context.Requirements.FlowParams.IsNewDataSet ||
+			req.Context.Requirements.FlowParams.ResourceMetadata != nil {
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:550]: This code is reached by either writing a new asset or updating an existing asset. In both cases it is problematic to use the supplied metadata.

Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go:550]: we will revisit this part. ResourceMetadata from flow requirements should be taken into account only when `isNewDataSet` is true

File: manager/controllers/app/fybrikapplication_controller.go (hunk 3 above)
Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go:568]: we will revisit this part. ResourceMetadata from flow requirements should be taken into account only when isNewDataSet is true

File: manager/controllers/app/data_catalog_interface.go (hunk 1 above)
Comment (revit13) [reply manager/controllers/app/data_catalog_interface.go:36]: will fix that thanks

File: manager/controllers/app/fybrikapplication_controller.go [hunk 5]
```diff
@@ -542,23 +546,26 @@ func (r *FybrikApplicationReconciler) checkGovernanceActions(configEvaluatorInpu
...
 			// update an existing dataset
 			// query the policy manager whether the operation is allowed
 			reqAction := policymanager.RequestAction{
 				ActionType:         configEvaluatorInput.Request.Usage,
 				Destination:        req.DataDetails.ResourceMetadata.Geography,
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:555]: req.DataDetails does not have Geography for a new asset.

File: manager/controllers/app/fybrikapplication_controller.go (hunk 3 above)
Comment (shlomitk1) [reply manager/controllers/app/fybrikapplication_controller.go:568]: Even not then. You don't know the destination at this point, so what will the WRITE query be about?

File: manager/controllers/app/fybrikapplication_controller.go (hunk 5 above)
Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go:555]: ok, it set to empty string in this case...

File: manager/controllers/app/fybrikapplication_controller.go (hunk 3 above)
Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go:568]: not sure if that answers the question but we have the tags so we can get policies related to it.

File: pkg/model/datacatalog/model.go [hunk 6]
```diff
@@ -43,3 +43,19 @@ type ResourceDetails struct {
 	// Data format
 	DataFormat taxonomy.DataFormat `json:"dataFormat,omitempty"`
 }
+
+// Return true if resourceMetadata is not empty. Otherwise return false.
+func ResourceMetadataNotEmpty(resourceMetadata *ResourceMetadata) bool {
```
[the remark this answers is not in the document]
Comment (revit13) [reply pkg/model/datacatalog/model.go]: Fixed thanks

**Comment by revit13** [comment]
@shlomitk1 please take a look at the updated version in particular the code related to the flow which calls policy compiler in `GetPoliciesDecisions`. please note that the policy compiler is been called also when `isNewDataSet` is true as we need to check for permission to write new data. it is still under testing but I appreciate your review. Thanks

File: charts/fybrik/files/taxonomy/fybrik_application.json [hunk 7]
```diff
@@ -41,11 +41,25 @@
...
+          "$ref": "#/definitions/FlowRequirements",
+          "description": "FlowParams include the requirements for particular data flows"
+        },
         "interface": {
           "$ref": "taxonomy.json#/definitions/Interface",
           "description": "Interface indicates the protocol and format expected by the data user"
         }
       }
+    },
+    "FlowRequirements": {
```
Comment (shlomitk1) [inline charts/fybrik/files/taxonomy/fybrik_application.json:54]: How was it generated? FlowRequirements has more fields than just metadata

Comment (revit13) [reply charts/fybrik/files/taxonomy/fybrik_application.json:54]: hmmm I ran `make all`... strange

**Comment by shlomitk1** [comment]
Looking now at the information sent to the policy manager, it seems that no changes have been made since updating the API. The purpose of Resource field in the request is to avoid an extra call to catalog connector, and separate the connectors completely. Why does opa-connector need to invoke GetAssetInfo if it receives the information about the resource in the request? 
Besides, it is not the responsibility of policy manager to decide how to merge two inputs of metadata.
Manager should provide to the policy manager connector the complete information about the asset in Resource field.

**Comment by revit13** [comment]
@shlomitk1 the call to GetAssetInfo from opa was before the resource metadata change. This pr contains the code to avoid it if resource metadata exists.

**Comment by shlomitk1** [comment]
> @shlomitk1 the call to GetAssetInfo from opa was before the resource metadata change. This pr contains the code to avoid it if resource metadata exists.

And if resource metadata does not exist? How would the call change anything? I'm saying that it is unnecessary, the calls to the policy manager should be updated accordingly.

File: manager/controllers/app/fybrikapplication_controller.go (hunk 3 above)
Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go:580]: @shlomitk1 as you suggested the metadata is now passed to the policy compiler and the call to the data catalog from policy compiler is removed.

File: manager/controllers/app/fybrikapplication_controller.go [hunk 8]
```diff
@@ -217,9 +218,22 @@ func (r *FybrikApplicationReconciler) checkReadiness(applicationContext Applicat
 				setErrorCondition(applicationContext, assetID, err.Error())
 				continue
 			}
+			reqResource := dataCtx.Requirements.FlowParams.ResourceMetadata
+			if reqResource != nil {
+				// we assume to have only the geography field set at this point
+				// in the provisionedBucket ResourceMetadata
+				geo := provisionedBucketRef.ResourceMetadata.Geography
+				if reqResource.Geography != "" && geo != reqResource.Geography {
+					// log an error and make a new attempt to register the asset
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:227]: Who will make a new attempt? Print warning that the geography field is going to be ignored, and continue with registration.

Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go]: Fixed. Thanks

File: manager/controllers/app/fybrikapplication_controller.go [hunk 9]
```diff
@@ -514,6 +528,9 @@ func (r *FybrikApplicationReconciler) constructDataInfo(req *DataInfo, appContex
 		}
 		logging.LogStructure("Catalog connector response", response, &log, zerolog.DebugLevel, false, false)
 		response.DeepCopyInto(req.DataDetails)
+	} else if req.Context.Requirements.FlowParams.ResourceMetadata != nil {
+		// Fill req.DataDetails with the metadata from the fybrikapplication
+		req.DataDetails.ResourceMetadata = *req.Context.Requirements.FlowParams.ResourceMetadata.DeepCopy()
```
Comment (shlomitk1) [inline manager/controllers/app/fybrikapplication_controller.go:533]: Why is DeepCopy needed?

Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go:533]: @shlomitk1 to avoid changing the values when it further processed.

Comment (shlomitk1) [reply manager/controllers/app/fybrikapplication_controller.go:533]: `req.DataDetails.ResourceMetadata = *req.Context.Requirements.FlowParams.ResourceMetadata` is enough, the structure is copied.

Comment (revit13) [reply manager/controllers/app/fybrikapplication_controller.go]: Fixed Thanks

**Review by shlomitk1 — APPROVED** [review]

**Pull request merged**